src/lib/pubkey/frodokem/frodokem.cpp

```diff
@@ -0,0 +1,398 @@
+            throw Invalid_Argument("FrodoKEM ciphertext does not have the correct byte count");
+         }
+
+         BufferSlicer ct_bs(encapsulated_key);
+         auto c_1 = ct_bs.take<FrodoPackedMatrix>(consts.len_packed_b_bytes());
+         auto c_2 = ct_bs.take<FrodoPackedMatrix>(consts.len_packed_c_bytes());
+         auto salt = ct_bs.take<FrodoSalt>(consts.len_salt_bytes());
+         BOTAN_ASSERT_NOMSG(ct_bs.empty());
+
+         const auto b_p = FrodoMatrix::unpack(consts, {consts.n_bar(), consts.n()}, c_1);
+         const auto c = FrodoMatrix::unpack(consts, {consts.n_bar(), consts.n_bar()}, c_2);
+
+         const auto w = FrodoMatrix::mul_bs(consts, b_p, m_private_key->s_trans());
+         const auto m = FrodoMatrix::sub(consts, c, w);
+
+         const auto seed_u_p = m.decode(consts);
+
+         shake.update(m_public_key->hash());
+         shake.update(seed_u_p);
+         shake.update(salt);
+
+         const auto seed_se_p = shake.output<FrodoSeedSE>(consts.len_se_bytes());
+         const auto k_p = shake.output<FrodoIntermediateSharedSecret>(consts.len_sec_bytes());
+         shake.clear();
+
+         shake.update(consts.encapsulation_domain_separator());
+         shake.update(seed_se_p);
+         const auto s_p =
+            FrodoMatrix::sample(consts,
+                                std::tuple(consts.n_bar(), consts.n()),
+                                shake.output<FrodoSampleR>(sizeof(uint16_t) * consts.n_bar() * consts.n()));
+
+         const auto e_p =
+            FrodoMatrix::sample(consts,
+                                std::tuple(consts.n_bar(), consts.n()),
+                                shake.output<FrodoSampleR>(sizeof(uint16_t) * consts.n_bar() * consts.n()));
+
+         auto b_pp = FrodoMatrix::mul_add_sa_plus_e(consts, s_p, e_p, m_public_key->seed_a());
+
+         const auto e_pp =
+            FrodoMatrix::sample(consts,
+                                std::tuple(consts.n_bar(), consts.n_bar()),
+                                shake.output<FrodoSampleR>(sizeof(uint16_t) * consts.n_bar() * consts.n_bar()));
+         shake.clear();
+
+         const auto v = FrodoMatrix::mul_add_sb_plus_e(consts, m_public_key->b(), s_p, e_pp);
+
+         const auto encoded = FrodoMatrix::encode(consts, seed_u_p);
+         auto c_p = FrodoMatrix::add(consts, v, encoded);
+
+         // The spec concats the matrices b_p and c (b_pp and c_p respectively)
+         // and performs a single CT comparison. For convenience we compare the
+         // matrices individually in CT and CT-&& the result.
+         //
+         // Note: b_p and c are unpacked values that are reduced by definition.
+         //       b_pp and c_p are calculated values that need the reduction for
+         //       an unambiguous comparison.
+         b_pp.reduce(consts);
+         c_p.reduce(consts);
+         const bool cmp_b_p_b_pp = b_p.constant_time_compare(b_pp);
+         const bool cmp_c_c_p = c.constant_time_compare(c_p);
```

Comment:
These `constant_time_compare` calls should probably return `CT::Mask` instead of bools

Reply:
Agreed, thanks for pointing this out!

`conditional_copy_mem` now also directly uses the mask instead of the value.